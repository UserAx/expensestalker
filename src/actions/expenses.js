import database from '../firebase/firebase';

// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

export const startAddExpense = (expenseData ={}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      description = '',
      note='',
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = {description, note, amount, createdAt};
    return database.ref(`users/${uid}/expenses`).push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    });
  }
}

export const startSetExpenses = ( ) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses`).once('value').then((snapshot) => {
      const expenses = [];
      snapshot.forEach((childSnapshot) => {
        expenses.push({id: childSnapshot.key,
                      ...childSnapshot.val()});
      });
      dispatch(setExpenses(expenses));
    });
  }
}

export const startRemoveExpense = ({id}={}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${id}`).remove().then(()=> {
      dispatch(removeExpense({id}));
    });
  }
}

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({//Note here we pass id as {id: 'someId'}
  type: 'REMOVE_EXPENSE',
  id
});

export const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`users/${uid}/expenses/${id}`).update(updates).then(()=> {
      dispatch(editExpense(id, updates));
    });
  }
}


// EDIT_EXPENSE
export const editExpense = (id, updates) => ({//However here we pass argument as ('someId', {description: 'Some Text'})
  type: 'EDIT_EXPENSE',
  id,
  updates
});