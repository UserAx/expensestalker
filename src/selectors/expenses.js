// Get visible expenses
import moment from 'moment';

export default (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses.filter((expense) => {//returns only those expense object that match the following cases.
    const createdAtMoment = moment(expense.createdAt);
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
    //Here startDate is getting compared to createdAtMoment not the other war around. 
    //if start day is yesterday and createdAtMoment is yesterday then we get true since startday is before today.
    //Same is true for my endDateMatch bellow.
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

    return startDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {//and sorts those object by amount or date if set.
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy === 'amount') {
      return a.amount < b.amount ? 1 : -1;
    }
  });
};