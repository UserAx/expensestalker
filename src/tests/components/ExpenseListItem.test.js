import React from 'react';
import ExpenseListItem from '../../components/ExpenseListItem';
import expenses from '../fixtures/expenses';
import {shallow} from 'enzyme';

test('should render expense list items', () => {
    const expense = {...expenses[0], id: 'staticIdforthistest'};
    //I have set a static id because every time the test runs the id is going to be unique i.e. it will fail every second time.
    const wrapper = shallow(<ExpenseListItem {...expense}/>);
    expect(wrapper).toMatchSnapshot();
});