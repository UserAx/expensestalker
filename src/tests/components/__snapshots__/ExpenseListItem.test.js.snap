// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense list items 1`] = `
<Link
  className="list-item"
  replace={false}
  to="/edit/staticIdforthistest"
>
  <div>
    <h3
      className="list-item__title"
    >
      rent
    </h3>
    <span
      className="list-item__sub-title"
    >
      January 1st, 1970
    </span>
  </div>
  <h3
    className="list-item__data"
  >
    $15.00
  </h3>
</Link>
`;
